restructuring the rise type system, adding lowering to imperative representation

martin-luecke opened the pull request:
Restructuring the rise type system + operations to better mix with other dialects.
Adding Lowering to affine+standard / scf+standard

martin-luecke committed:
added current state of conversion experiments

martin-luecke committed:
Added riseModule op + matching against it for rise to imperative lowering

martin-luecke committed:
detecting an application of reduction and extraction of corresponding information

martin-luecke committed:
sync

martin-luecke committed:
transformation is now valid, but not emitting the transformed IR

martin-luecke committed:
merge shape Dialect

martin-luecke committed:
fixing Attributes

martin-luecke committed:
switching to changed dialect and pass registration

martin-luecke committed:
more conversion experiments

martin-luecke committed:
valid identity conversion

martin-luecke committed:
sync experiments

martin-luecke committed:
basic transformations working now, fixed TableGen Target

martin-luecke committed:
adopting clang-format style

martin-luecke committed:
started writing a lowering strategy for reduce

martin-luecke committed:
lowering of a simple reduction to llvm

martin-luecke committed:
fixed integration with check-mlir

martin-luecke committed:
rise.module -> rise.fun

martin-luecke committed:
added possibility to use the result of a rise computation afterwards.

martin-luecke committed:
removing Types from dialect registration, which should never actually appear in the IR (DataType)

martin-luecke committed:
started working on lowering a map to imperative

martin-luecke committed:
added support for lowering when applying a function in multiple partial applies.

martin-luecke committed:
working on Lambda Lowering

martin-luecke committed:
further expiremnts for lowering lambdas

martin-luecke committed:
work on lowering lambda

martin-luecke committed:
rethinking lowering of Lambda

martin-luecke committed:
proper lowering of map...lambda, restructuring lowering

martin-luecke committed:
problem with lowering literals

martin-luecke committed:
work on lowering lambda

martin-luecke committed:
rise.add and rise.mult can now be expanded to Lambdas. Fixed a bug in the parsing of Lambdas

martin-luecke committed:
starting with OutputPath

martin-luecke committed:
working on parsing for rise.fun, adding std::variant header file

martin-luecke committed:
changing interface of rise.fun, adding rise.in, changing lowering accordingly

martin-luecke committed:
changing structure of rise.fun

martin-luecke committed:
working on rise.idx

Commit:
introducing intermediate patterns

Commit:
basic compilation works using intermediate patterns. Still ad-hoc path resolution.

Commit:
lowering works for map_map_add

Commit:
working on zip lowering

Commit:
can generate code for zip, working on reduce

Commit:
improving lowering of reduce

Commit:
lowering reduce to imperative works, fixing FileCheck tests.

Commit:
lowering dot to imperative

Commit:
basic lowering for mm

Commit:
rise.fun can now take input arguments. Support for casting input to Datatypes via rise.in.

Commit:
removed hardcoded lowering of literals.

Commit:
rise.mult |-> rise.mul

Commit:
work on parsing literals

Commit:
forgot to add header file

Commit:
modifying mlir Parser expose the parsing of lists (look exactly like our arrays)

Commit:
rise.reduce |-> rise.reduceSeq. rise.map divided into mapSeq and mapPar. Experiments with lowering mapPar to affine.for

Commit:
added option to lower MapSeq, MapPar, ReduceSeq to different target dialects using Attribute

Commit:
merge master

Commit:
removing obsolete riseToStandard pass, cleanup

Commit:
cleanup

Commit:
Revert "cleanup"

This reverts commit e1a911d87694e3e3101e3d1f4aee81feb1fc493e.

Commit:
delete riseToStandard leftovers

Commit:
remove unused include

Commit:
adding type rise.scalar with parsing & printing support

Commit:
added rise.wrap & rise.unwrap operations + parsing

Commit:
fixed bug in unsafe SSA lookup

Commit:
rewriting parsing completely, adjusting Ops and Attributes to support scalar type. Changing tests to use scalar type

Commit:
lowering working for simple examples.

Commit:
lowering works using rise.scalar wrapping with combination with other dialects

Commit:
prepare for moving away from rise.fun

Commit:
moving lowering pass to operate on FuncOp

Commit:
sync

Commit:
moving tests to new structure

Commit:
fixed all tests

Commit:
pulling in new changes

Commit:
Merge branch 'master', commit '57d8b8d6f0b91b380ca3b270b4439c338ed67f53' of https://github.com/llvm/llvm-project into feature/relaxedTypesystem

Commit:
introduced rise.embed, removed rise.wrap & rise.unwrap, adjusted compilation accordingly.

Commit:
Merge branch 'master', commit 'c1ae72d03f71bb0a4cf7c39ce5ffe3dbc8b96514' of https://github.com/llvm/llvm-project into feature/relaxedTypesystem

Commit:
adjusting parsing of rise.lambda

Commit:
adding parsing of minus token to mlir core

Commit:
added rise.out operation and changed lowering accordingly

Commit:
cleanup of lowering to imperative

Commit:
Merge commit '478f2ce5d3c44565b328c18d999deb6d1f3c7ee1' of github.com:llvm/llvm-project into feature/relaxedTypesystem

Commit:
intermediate codegen representation now based on rise type system, introduced rise.codegen.cast

Commit:
fixing bugs in lowering

Commit:
Merge commit '49cbe56a657b91e612f8305e7f8f9119ffe84378' of github.com:llvm/llvm-project into feature/relaxedTypesystem

Commit:
enabling codegen, moving to llvm commit 49cbe56 to explore integration with iree

Commit:
merge master

Commit:
cleanup, preparing for PR

Commit:
Merge commit 'c4b5a66e44f031eb89c9d6ea32b144f1169bdbae' of github.com:llvm/llvm-project into feature/relaxedTypesystem

Commit:
remove dependency on outside BlockArgs for our translation

Commit:
added irregular mm test, fixed indexing bug, removed RISE prefix of intermediate operations

Commit:
adding EDSC for mapSeq

Commit:
work on edsc builders

Commit:
expanding builders

Commit:
expanding Rise EDSC

Commit:
fixing builder of embed

Commit:
parsing for split, join, slide, pad + example for a 1D stencil

michel-steuwer commented on README.md:
```diff
@@ -1,110 +1,13 @@
-This directory and its sub-directories contain source code for LLVM,
-a toolkit for the construction of highly optimized compilers,
-optimizers, and run-time environments.
+Check out our current implementation in the branch [feature/relaxedTypesystem](https://github.com/rise-lang/mlir/tree/feature/relaxedTypesystem)
```
Maybe you want to update this to reflect the correct link for after the merge.

martin-luecke replied:
Updated README in 6de2b37931530e4de2a4e1a58e81ba7fea60179b

michel-steuwer commented on mlir/lib/Transforms/DialectConversion.cpp:
```diff
@@ -1402,8 +1402,8 @@ OperationLegalizer::legalizePattern(Operation *op, RewritePattern *pattern,
-  assert((replacedRoot || updatedRootInPlace()) &&
-         "expected pattern to replace the root operation");
+//  assert((replacedRoot || updatedRootInPlace()) &&
+//         "expected pattern to replace the root operation");
```
Can you comment on why this assertion can be removed, or if this was only temporary and the assertion should be reintroduced?

martin-luecke replied:
This was used for local testing. Assertion reintroduced in b90b46b7eb9f9d3a081630970b2d5e2764e719f6.

michel-steuwer commented on mlir/lib/Dialect/Rise/IR/Types.cpp:
```diff
@@ -0,0 +1,126 @@
+  return Base::get(context, RiseTypeKind::RISE_ARRAY, size, elementType);
+}
+
+//mlir::LogicalResult
```
Can this be removed?

martin-luecke replied:
removed :+1:

michel-steuwer commented on mlir/lib/Dialect/Rise/IR/RiseDialect.cpp:
```diff
@@ -0,0 +1,204 @@
+
+    Type input;
+    Type output;
+    if (parser.parseLess())
```
This reads as if when `parser.parseLess` is successful we will return `nullptr` where actually the opposite is true.

Can the `if` statement be written in a way to make clear that on failure the branch is taken?

martin-luecke replied:
Clarified parsing in b90b46b7eb9f9d3a081630970b2d5e2764e719f6

michel-steuwer commented on mlir/lib/Dialect/Rise/IR/RiseDialect.cpp:
```diff
@@ -0,0 +1,204 @@
+
+    Type wrappedType;
+    if (parser.parseLess() || parser.parseType(wrappedType) ||
+        parser.parseGreater())
```
Similarly here, it seems odd to me to use here `or` and not `and` to compose the parsing.

martin-luecke replied:
Clarified parsing in b90b46b7eb9f9d3a081630970b2d5e2764e719f6

michel-steuwer commented on mlir/include/mlir/Dialect/Rise/IR/Rise.td:
```diff
@@ -0,0 +1,361 @@
+def DataType
+    : Type<
+          Or<[TupleType.predicate, ArrayType.predicate, ScalarType.predicate]>>;
+def NatOrDataType : Type<Or<[NatType.predicate, DataType.predicate]>>;
```
Where do you need this?

martin-luecke replied:
In this same file this is used to describe invariants on the operations. Example: rise.in always has to return a DataType

michel-steuwer commented on mlir/include/mlir/Dialect/Rise/IR/Rise.td:
```diff
@@ -0,0 +1,361 @@
+
+def InOp : Rise_Op<"in"> {
+  let summary = "rise in";
+  let description = "";
```
We should provide a (brief) description of every Op.

martin-luecke replied:
Introduced description of every Op in 6de2b37931530e4de2a4e1a58e81ba7fea60179b

michel-steuwer commented on mlir/include/mlir/Dialect/Rise/IR/Rise.td:
```diff
@@ -0,0 +1,361 @@
+
+def EmbedOp : Rise_Op<"embed", [IsolatedFromAbove]> {
+  let summary = "rise embed";
+  let arguments = (ins Variadic<ScalarType> : $exposedValues);  // This is not really used. We have region arguments, not "normal" arguments.
```
What does this comment suggest?

What does `"not really used"` mean here?

martin-luecke replied:
Comment is left over from when I originally implemented this. exposedValues is actually required, as this operation has arguments of type scalar _and_ blockArguments of the wrapped type.
Comment is already removed

michel-steuwer commented on mlir/include/mlir/Dialect/Rise/IR/Rise.td:
```diff
@@ -0,0 +1,361 @@
+  let description =
+      "This operation is used as a terminator for the region of a lambda and "
+      "embed. "
+      "We will experiment with using the standard return operation instead.";
```
Have we actually concluded something about the `rise.return` vs. `std.return`?

martin-luecke replied:
We have, the std.return can not be used as a return in our lambda or embed. Hence we use rise.return as a terminator. This is done similarly in other dialects e.g. in the gpu dialect, where gpu.return is used as a terminator for gpu.func

michel-steuwer commented on mlir/include/mlir/Dialect/Rise/IR/Rise.td:
```diff
@@ -0,0 +1,361 @@
+def MapSeqOp : Rise_Op<"mapSeq"> {
+  let summary = "mapSeq";
+  let description =
+      "This operation returns a function which expects a function from s to t "
```
Does it make sense to say `"This operation returns ..."` or shouldn't it be `"This operation represents ..."`?

michel-steuwer replied:
I would also say that you want to distinguish the *operation* (i.e. `map`) from the *function* that is the argument passed to it.

I would suggest here something like:
```
The mapSeq operation expects a function from s to t and an array of length n with elements of type s. The function is to be applied sequentially to all elements of the array.
```

martin-luecke replied:
Well the problem is this operation doesn't expect a function. The application of it does. I think it is important to stress that this has to be applied explicitly.
I guess we could say:
`The mapSeq operation represents the functional mapSeq pattern. Its application expects a function from s to t and an array of length n with elements of type s. The function is to be applied sequentially to all elements of the array.`

martin-luecke replied:
> Does it make sense to say `"This operation returns ..."` or shouldn't it be `"This operation represents ..."`?

Actually it literally returns that function

michel-steuwer commented on mlir/include/mlir/Dialect/Rise/IR/Rise.td:
```diff
@@ -0,0 +1,361 @@
+}
+
+def SplitOp : Rise_Op<"split"> {
+  let summary = "split";
```
For a description I would suggest something like:
```
The split operation expects a natural number n an array of length m and returns a nested array of length m/n where every element is an array of size n. The value of the elements of the input array remains the same.
```

michel-steuwer commented on mlir/include/mlir/Dialect/Rise/IR/Rise.td:
```diff
@@ -0,0 +1,361 @@
+//===----------------------------------------------------------------------===//
+
+def IdxOp : Rise_Op<"codegen.idx"> {
+  let summary = "rise idx";
```
These also deserve a (brief) description.

martin-luecke replied:
Description of all the ops in 6de2b37931530e4de2a4e1a58e81ba7fea60179b

michel-steuwer commented on mlir/include/mlir/Dialect/Rise/IR/Rise.td:
```diff
@@ -0,0 +1,361 @@
+}
+
+//===----------------------------------------------------------------------===//
+// Arithmetic operations
```
Do we still need them?

martin-luecke replied:
I still had some tests depending on them. I removed all leftovers here and in the lowering

michel-steuwer commented on mlir/include/mlir/Conversion/RiseToImperative/ConvertRiseToImperative.h:
```diff
@@ -0,0 +1,57 @@
+
+using OutputPathType = mpark::variant<int, Value>;
+
+void AccT(ApplyOp apply, Value out, PatternRewriter &rewriter);
```
Does it make sense to have a single `AccT` with the first argument being a `Value` and then dispatch internally?
I guess this would lose some of the static typing guarantees you get by having the two versions.

martin-luecke replied:
changed to a single AccT (actually now lowerAndStore) function :+1:

michel-steuwer commented on mlir/include/mlir/Conversion/RiseToImperative/ConvertRiseToImperative.h:
```diff
@@ -0,0 +1,57 @@
+                          PatternRewriter &rewriter);
+void generateReadAccess(SmallVector<OutputPathType, 10> path, Value storeVal,
+                        Value storeLoc, PatternRewriter &rewriter);
+
```
Do you need to make all of these functions public in the header?

If you only use them locally in the file then this would not be required and result in a cleaner internface.

martin-luecke replied:
moved

michel-steuwer commented on mlir/include/mlir/Dialect/Rise/variant.hpp:
```diff
@@ -0,0 +1,2813 @@
+// MPark.Variant
```
Is there not a variant type somewhere in the MLIR/LLVM utils implementation?

michel-steuwer replied:
I guess that including this could have implications for the MLIR/LLVM license and might prevent this to be merged upstream ...

martin-luecke replied:
I agree, as discussed we will leave it in for now.

michel-steuwer commented on mlir/include/mlir/IR/OpImplementation.h:
```diff
@@ -497,6 +500,10 @@ class OpAsmParser {
                                      SmallVectorImpl<Value> &result) = 0;
 
+  /// Resolve an operand to an SSA value
+  virtual ParseResult resolveOperandUnsafe(const OperandType &operand,
```
Is `unsafe` really the right name here?
Maybe it is because it echos some internal API, but at this level it seems that this is just asking to resolve the operand without the type.

martin-luecke replied:
I renamed this to just `resolveOperand`

michel-steuwer commented on mlir/lib/Parser/Parser.cpp:
```diff
@@ -2296,6 +2312,7 @@ ParseResult TensorLiteralParser::parseList(SmallVectorImpl<int64_t> &dims) {
   // Return the sublists' dimensions with 'size' prepended.
   dims.clear();
   dims.push_back(size);
+  std::cout << "size pushed back!: " << size << "\n" << std::flush;
```
This seem to be a debug output!?

martin-luecke replied:
Indeed, removed.

michel-steuwer commented on mlir/lib/Parser/Parser.cpp:
```diff
@@ -3725,6 +3758,20 @@ Value OperationParser::resolveSSAUse(SSAUseInfo useInfo, Type type) {
 
+/// Given an unbound reference to an SSA value and its type, return the value
+/// it specifies.  This returns null on failure.
+Value OperationParser::resolveSSAUseUnsafe(SSAUseInfo useInfo) {
```
The implementation doesn't look unsafe to me ...

martin-luecke replied:
this is now just `resolveSSAUse`

Commit:
added lowering for join and split. Possible bug in SplitAcc,JoinAcc

michel-steuwer commented on mlir/lib/Dialect/Rise/IR/Ops.cpp:
```diff
@@ -0,0 +1,904 @@
+  Type outputType;
+
+  SmallVector<OpAsmParser::OperandType, 4> arguments;
+  SmallVector<Type, 4> argumentTypes = SmallVector<Type, 4>();
```
You do not have to call the default constructor explicitly for local stack variables, i.e.
```
SmallVector<Type, 4> argumentsTypes;
```
is here the same as
```
SmallVector<Type, 4> argumentsTypes = SmallVector<Type, 4>();
```

See: https://en.cppreference.com/w/cpp/language/default_initialization

martin-luecke replied:
:+1: These are leftovers when I started using SmallVectors and experimented with them

michel-steuwer commented on mlir/lib/Dialect/Rise/IR/Ops.cpp:
```diff
@@ -0,0 +1,904 @@
+  if (parser.parseLParen())
+    return failure();
+
+  // Not working correct
```
Does this need to be fixed or can this be removed?

martin-luecke replied:
We don't have rise.fun anymore, so I removed the parsing for it :+1:

michel-steuwer commented on mlir/lib/Dialect/Rise/IR/Ops.cpp:
```diff
@@ -0,0 +1,904 @@
+      return failure();
+  }
+
+  // TODO: we dont want to have to explicitly give our type
```
Can this now be refactored here (and the local variable `simplified` be removed)?

martin-luecke replied:
:+1:

michel-steuwer commented on mlir/lib/Dialect/Rise/IR/Ops.cpp:
```diff
@@ -0,0 +1,904 @@
+    failure();
+
+  result.addTypes(funType.getOutput());
+  //  result.setOperandListToResizable(true);
```
Remove the comment?

martin-luecke replied:
:+1:

Commit:
lowering of slide + example in stencil.mlir, improving builder for lambdaOP

Commit:
builder for EmbedOp

Commit:
builders for constructing types

Commit:
lowering for split, join, pad

Commit:
Merge branch 'feature/relaxedTypesystem' of github.com:rise-lang/mlir into feature/relaxedTypesystem

Commit:
fixing tests

Commit:
improving EDSC builders

Commit:
refactoring ArrayType and Tuple to contain DataTypes rather than Types.

Commit:
extending builders, introducing slide2d higher level abstraction

Commit:
split & join builders progress

martin-luecke committed:
builders for split & join, introducing mapOp

Commit:
builders for lambda with fixed number of arguments

Commit:
extending builders

michel-steuwer commented on mlir/lib/Dialect/Rise/IR/Ops.cpp:
```diff
@@ -0,0 +1,904 @@
+  // parsing the optional attribute specifying a lowering target
+  SmallVector<std::string, 4> loweringTargets = {"affine", "loop"};
+  NamedAttrList attributesFromDict;
+  if (succeeded(parser.parseOptionalAttrDict(attributesFromDict))) {
```
Can this complicated parsing be factored out into an individual function?

martin-luecke replied:
:+1:

michel-steuwer commented on mlir/lib/Dialect/Rise/IR/Ops.cpp:
```diff
@@ -0,0 +1,904 @@
+  // parsing the optional attribute specifying a lowering target
+  SmallVector<std::string, 4> loweringTargets = {"affine", "loop"};
+  NamedAttrList attributesFromDict;
+  if (succeeded(parser.parseOptionalAttrDict(attributesFromDict))) {
```
This seems to be (almost?) exactly the same as above ... factoring this out into a function seems beneficial.

martin-luecke replied:
Will be in next commit

michel-steuwer commented on mlir/lib/Dialect/Rise/IR/Ops.cpp:
```diff
@@ -0,0 +1,904 @@
+    failure();
+
+  result.addTypes(
+      FunType::get(builder.getContext(),
```
It might make sense to use the simplified type builder API here (now that you have it).

martin-luecke replied:
The builders require there to be a ScopedContext. As this is called from outside our code, we'd have to initialize one here and I couldn't figure out how to do it here.

martin-luecke replied:
Nevermind, I figured it out.

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+  });
+
+  // Start at the back and find the rise.out op
+  OutOp outOp;
```
You could write this using the `std::find_if` algorithm as something like:                                                                                                                  
```                                                                                                                                                                                         
auto outOp = cast<OutOp>(* std::find_if(block.rbegin(), block.rend(), [](auto op) { return isa<OutOp>(*op); } ) );                                                                          
```

michel-steuwer replied:
I would generally suggest using this style as it is expressing the intend (finding) clearly (and avoids the possibility to introduce bugs in the low-level loop code).

martin-luecke replied:
:+1:

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+        isa<MapParOp>(inst) || isa<ReduceSeqOp>(inst) || isa<OutOp>(inst) ||
+        isa<LiteralOp>(inst) || isa<TransposeOp>(inst) || isa<SplitOp>(inst) ||
+        isa<JoinOp>(inst) || isa<SlideOp>(inst) || isa<PadOp>(inst)) {
+      if (inst->getParentOfType<LambdaOp>()) {
```
I would suggest negating the predicate and removing the `else` branch.

martin-luecke replied:
:+1:

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+    return;
+  });
+  size_t unneededLeftoverOps = leftoverOps.size();
+  for (size_t i = 0; i < unneededLeftoverOps; i++) {
```
can't you express this as a `while( auto op = pop_back(); ) { .. }` loop?

martin-luecke replied:
I opted for 
```
  while (!leftoverOps.empty()) {
    auto op = leftoverOps.pop_back_val();
    ...
 }
```
SmallVector does not like pop_back_val() when it is empty.

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+  funcOp.walk([&assignOps](Operation *inst) {
+    if (!inst)
+      return;
+    if (isa<AssignOp>(inst)) {
```
I would suggest combining the conditions here into a single if: `if (inst && isa<AssignOp>(inst)) { ... }`

martin-luecke replied:
:+1:

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+
+  funcOp.dump();
+  // Codegen:
+  bool doCodegen = true;
```
Is this `doCodegen` variable a debug artifact? Can we remove it?

martin-luecke replied:
This is the trigger to enable or disable the second phase of the lowering. This is very useful for emitting our codegen representation. I'd argue for keeping it in for now and at some point divide the two lowering phases into two passes.

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+    funcOp.walk([&](Operation *inst) {
+      if (!inst->getDialect())
+        return;
+      if (inst->getDialect()->getNamespace().equals("rise")) {
```
Same as before ... combine the `if`s

martin-luecke replied:
`if (inst->getDialect() && inst->getDialect()->getNamespace().equals("rise"))` only works when after evaluating the lhs of && to false the rhs is not evaluated, as that would be dereferencing a nullptr. I wasn't sure about this, but just looked it up and it is in the standard :+1:

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+  }
+
+  size_t unneededOps = erasureList.size();
+  for (size_t i = 0; i < unneededOps; i++) {
```
Same as before: use a `while(pop) ` loop

martin-luecke replied:
:+1:

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+  return;
+}
+
+/// Acceptor Translation
```
Maybe we can call this `translateOpAndStoreResult` and the continuation translation `translateOp`.

Alternatives:
 - `transOp` + `transOpAndStoreRes`
 - `transOp` + `transOpAndStore`
 - `gen` + `genAndStoreRes`
 - `gen` + `genAndStore`

martin-luecke replied:
I opted for `lower` + `lowerAndStore`

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+          dyn_cast<ApplyOp>(returnOp.getOperand(0).getDefiningOp())) {
+    AccT(apply, out, rewriter);
+    return;
+  } else if (EmbedOp embedOp =
```
I would suggest writing this as a separate `if` (the `return` in the branch before makes the `else` unnecessary)

martin-luecke replied:
:+1:

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+    //    rewriter.eraseOp(embedOp);
+
+    return;
+  } else {
```
I would suggest removing the `else` altogether and to the check outside any branching.

martin-luecke replied:
:+1:

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+
+    auto assignment = rewriter.create<AssignOp>(newEmbed->getLoc(),
+                                                newEmbed->getResult(0), out);
+    //    rewriter.eraseOp(embedOp);
```
remove comment?

martin-luecke replied:
:+1:

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+  Operation *appliedFun = apply.getOperand(0).getDefiningOp();
+  Location loc = apply.getLoc();
+
+  if (isa<ReduceSeqOp>(appliedFun)) {
```
I would strongly suggest putting these branches into individual functions.
They can (and should) be local to this file, but would greatly increase readability!

michel-steuwer replied:
I imagine something like:
```
if (isa<ReduceSeqOp>(appliedFun)) {
	emitRemark(appliedFun->getLoc()) << "AccT of ReduceSeq";
	reduceSeqTranslateOpAndStoreResult(
      appliedFun->getAttrOfType<NatAttr>("n"),
      appliedFun->getAttrOfType<DataTypeAttr>("s"),
      appliedFun->getAttrOfType<DataTypeAttr>("t"),
      apply.getOperand(1),
      apply.getOperand(2),
      apply.getOperand(3)
    );
}
```

martin-luecke replied:
done :+1:

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+
+    bool defineNewAccumulator = false;
+    // Accumulator for Reduction
+    // TODO: do this properly! This can be way better structured
```
This seems to be a bit in flux.
Can we expand on the comment a bit more or remove it?

martin-luecke replied:
cleaned up

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+      rewriter.setInsertionPointAfter(embedOp);
+      accum = embedOp.getResult();
+    } else {
+      // TODO: not sure about this one
```
Same here: explain the questions/problems a bit more or remove the comment.

martin-luecke replied:
removed

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+        indexedAccum = accum;
+      }
+
+      //      EmbedOp embedOp = rewriter.create<EmbedOp>(
```
Remove?

martin-luecke replied:
removed

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+      auto initAccum = rewriter.create<AssignOp>(appliedFun->getLoc(), contInit,
+                                                 indexedAccum);
+    }
+    // zero constant for indexing
```
Remove?

martin-luecke replied:
removed

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+    Block *forLoopBody;
+
+    // lowering to a specific loop depending on the lowering target dialect
+    std::string loweringTarget;
```
With a lambda you can avoid the empty initialization of the string (and naming it three times in the following code):
```C++
std::string loweringTarget = [&]{
  if (StringAttr loweringTargetAttr =
            appliedFun->getAttrOfType<StringAttr>("to")) {
      return loweringTargetAttr.getValue().str();
  } else {
      return "loop"; // default lowering target
  }
}();
```
(not that the lambda is called in the last line)

martin-luecke replied:
Very nice!

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+    rewriter.setInsertionPointToStart(forLoopBody);
+    LambdaOp reductionLambda = dyn_cast<LambdaOp>(reductionFun.getDefiningOp());
+
+    // TODO: not needed anymore
```
Clean up the comments below?

martin-luecke replied:
cleaned up

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+    //      //      std::cout << "contArray is not of MemrefType";
+    //    }
+
+    if (contArray.getType().isa<ArrayType>()) {
```
Isn't this a requirement by the type of `reduceSeq`? If so this should be an assertion not a branch where if this is not true `xi` is uninitiallized.

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+          contArray, loopInductionVar);
+    }
+
+    // I think we don't need these anymore. Just pass the result of rise.embed
```
Clean up the comments.

martin-luecke replied:
removed

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+                                        lambdaCopy.getResult(),
+                                        ValueRange{accum, xi.getResult()});
+
+    AccT(fxi, accum, rewriter);
```
Maybe add a comment before this call to explain what is generated/translated here.

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+    }
+
+    if (loweringTarget == "loop") {
+      auto lowerBound =
```
Maybe you can write a convenient function local to this file, that creates a `mlir::scf::ForOp` from our own types (as this is the same code as for the `reduceSeq` op).

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+
+    IdxOp xi;
+
+    // TODO: not needed anymore
```
Clarify the comment or adjust the code.

martin-luecke replied:
removed

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+
+      xi = rewriter.create<IdxOp>(loc, inIndexOpResult, contArray,
+                                  loopInductionVar);
+    } else if (isa<IdxOp>(contArray.getDefiningOp())) {
```
Can this empty branch be removed?

martin-luecke replied:
removed

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+      //      std::cout << "got an idx and not a memref! \n" << std::flush;
+    }
+
+    if (contArray.getType().isa<ArrayType>()) {
```
Same comment as for `ReduceSeq`. This should be an assertion.

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+    auto fxi = rewriter.create<ApplyOp>(loc, lambdaCopy.getType(),
+                                        lambdaCopy.getResult(), xi.getResult());
+
+    //    ArrayRef<int64_t> outShape =
```
Remove comment.

martin-luecke replied:
removed

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+
+    rewriter.setInsertionPointToStart(forLoopBody);
+
+    LambdaOp fLambda = dyn_cast<LambdaOp>(f.getDefiningOp());
```
Clear the comments in the following lines.

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+    //      //      std::cout << "got an idx and not a memref! \n" <<
+    //      std::flush;
+    //    }
+    if (contArray.getType().isa<ArrayType>()) {
```
=> assertion

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+           splitAccInterm.getResult(), rewriter);
+    } else {
+      emitError(appliedFun->getLoc())
+          << "input to Split has to be result of ApplyOp"; // Is this actually a
```
I guess the other case (besides an `ApplyOp`) could be an `rise.in` op.

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+
+    // Find return in Lambda Region to start new AccT
+    rise::ReturnOp returnOp;
+    for (auto op = lambda.region().front().rbegin();
```
Use `std::find_if`.

martin-luecke replied:
I used `rise::ReturnOp returnOp = dyn_cast<ReturnOp>(lambda.region().front().getTerminator());`

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+                                                newMulOp.getResult(), out);
+
+    return;
+  } else {
```
I would (again) suggest a code style of:
```
if (...) {
  ...
  return;
}
if (...) {
  ...
  return;
}
// default case 
```
over:
```
if (...) {
  ...
  return;
} else if (...) {
  ...
  return;
} else {
  // default case
}
```

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+  }
+}
+
+/// Continuation Translation
```
see comment at `AccT` for naming suggestions.

martin-luecke replied:
done

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+  Location loc = contValue.getLoc();
+  auto oldInsertPoint = rewriter.saveInsertionPoint();
+
+  if (contValue.isa<OpResult>()) {
```
I would suggest negating the predicate and exit early to avoid the additional (unecessary) level of indentation.

martin-luecke replied:
done

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+  auto oldInsertPoint = rewriter.saveInsertionPoint();
+
+  if (contValue.isa<OpResult>()) {
+    if (isa<LiteralOp>(contValue.getDefiningOp())) {
```
I would (as for `AccT`) suggest individual functions for each Op

martin-luecke replied:
done

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+
+      emitRemark(contValue.getLoc()) << "Literal value: " << literalValue;
+
+      if (LiteralOp op = dyn_cast<LiteralOp>(contValue.getDefiningOp())) {
```
I would (again) suggest an early exit to avoid the additional level of indentation.

martin-luecke replied:
:+1:

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+            shape.push_back(arrayType.getSize().getIntValue());
+          }
+
+          Type memrefElementType = FloatType::getF32(rewriter.getContext());
```
This currently restricts the arrays to have an element type of F32 (specifically this forbids nested arrays).
This should be clearly stated as a limitation of this current implementation - either as a comment, or even better with an assertion and a comment.

martin-luecke replied:
It is true that the element type is restricted to F32, however this does not hinder the shape having an arbitrary number of entries. e.g memref<2x4x6xf32> (shape is [2,4,6])

Anyway, I fixed it

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+        auto contLhs = ConT(lhs, rewriter.getInsertionPoint(), rewriter);
+        auto contRhs = ConT(rhs, rewriter.getInsertionPoint(), rewriter);
+
+        // usually this is an Array of tuples. But at the end it always has to
```
This comment should hopefully no longer apply and be removed.

martin-luecke replied:
correct :+1:

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+                                        tmpArray.getResult());
+
+        rewriter.setInsertionPointAfter(embedOp);
+        // TODO: This is not correct I think!
```
Either elaborate or remove the comment.

martin-luecke replied:
removed

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+    emitError(rewriter.getUnknownLoc()) << "codegen started with nullptr!";
+  }
+  if (AssignOp assign = dyn_cast<AssignOp>(op)) {
+    emitRemark(op->getLoc()) << "Codegen for Assign";
```
I would prefer an early exit.

martin-luecke replied:
:+1:

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+mlir::rise::codeGenStore(Value storeLocation, Value val,
+                         SmallVector<OutputPathType, 10> path,
+                         PatternRewriter &rewriter) {
+  if (storeLocation.isa<OpResult>()) {
```
Early exit.

martin-luecke replied:
restructured :+1:

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+Value mlir::rise::codeGen(Value val, SmallVector<OutputPathType, 10> path,
+                          PatternRewriter &rewriter) {
+
+  if (val.isa<OpResult>()) {
```
early exit.

martin-luecke replied:
restructured :+1:

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+      Value cst1 = rewriter.create<ConstantIndexOp>(loc, 1).getResult();
+
+      Value n_minus_1 = rewriter.create<SubIOp>(loc, n, cst1).getResult();
+//      Value n_minus_r_minus_1 = rewriter.create<SubIOp>(loc, n_minus_r, cst1).getResult();
```
remove comment

martin-luecke replied:
:+1:

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+
+      path.push_back(ifthenVal);
+
+      // TODO: bronching!
```
comment?

martin-luecke replied:
:+1:

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+      indexValues.push_back(*val);
+    }
+  }
+  //  // handle problem originating from translation of reduce (accessing
```
remove comment

martin-luecke replied:
:+1:

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+  SmallVector<Value, 10> indexValues = {};
+  for (auto element = path.rbegin(); element != path.rend(); ++element) {
+    if (auto i = mpark::get_if<int>(&*element)) {
+      index = *i;
```
Is it deliberate that `index` might be overridden multiple times?

martin-luecke replied:
Index is actually never used here. This is a leftover from when it was not 100% clear what the path looks like. I removed that part

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+
+  ConversionTarget target(getContext());
+
+  //  target.addLegalOp<
```
remove comment?

martin-luecke replied:
:+1:

michel-steuwer commented on mlir/lib/Conversion/RiseToImperative/ConvertRiseToImperative.cpp:
```diff
@@ -0,0 +1,1638 @@
+    return !riseInside;
+  });
+
+  //  if (failed(applyPartialConversion(module, target, patterns)))
```
remove comment?

martin-luecke replied:
:+1:

michel-steuwer commented on mlir/lib/Dialect/Rise/EDSC/Builders.cpp:
```diff
@@ -0,0 +1,665 @@
+                               ValueRange{lambda, initializer, array});
+}
+
+// something fishy here! Type is not correct. It is scalar -> scalar instead of
```
Is this comment still relevant? If so, we should investigate.

martin-luecke replied:
Comment is left from a bigger debugging session. Will be removed in next commit

michel-steuwer commented on mlir/test/Conversion/RiseToImperative/mm_fused.mlir:
```diff
@@ -0,0 +1,114 @@
+        }
+    }
+
+//    %A = alloc() : memref<4x4xf32>
```
can this be cleaned up?

martin-luecke replied:
:+1:

michel-steuwer commented on mlir/test/Conversion/RiseToImperative/mm_irregular.mlir:
```diff
@@ -0,0 +1,147 @@
+        }
+    }
+
+//    %A = alloc() : memref<4x4xf32>
```
cleanup?

martin-luecke replied:
:+1:

michel-steuwer commented on mlir/test/Conversion/RiseToImperative/pad.mlir:
```diff
@@ -0,0 +1,61 @@
+
+    %print_me_in = memref_cast %inputArray : memref<9xf32> to memref<*xf32>
+    %print_me_out = memref_cast %outputArray : memref<14xf32> to memref<*xf32>
+//    call @print_memref_f32(%print_me_in): (memref<*xf32>) -> ()
```
remove this comment?

martin-luecke replied:
:+1:

michel-steuwer commented on mlir/test/Conversion/RiseToImperative/stencil.mlir:
```diff
@@ -0,0 +1,68 @@
+
+    %print_me_in = memref_cast %inputArray : memref<9xf32> to memref<*xf32>
+    %print_me_out = memref_cast %outputArray : memref<11xf32> to memref<*xf32>
+//    call @print_memref_f32(%print_me_in): (memref<*xf32>) -> ()
```
can be removed?

martin-luecke replied:
:+1:

michel-steuwer reviewed (changes requested):
Overall in a very good state 👍
Here are lot's of small suggestions for improvements for more clear and maintainable code.

Commit:
cleanup of rise EDSC, start work on lowering map

Commit:
Addressing PR comments:

- clarifying parsing syntax
- cleanup of modifications to mlir parser
- removing debug prints
- removing clutter in tests

Commit:
Support lowering of map operation

Commit:
moving generation of rise programs from iree fork to EDSC/highlevel

Commit:
expanding builders (pad2D, mapSeq2D), fixing literal parsing bug, tests for builders

Commit:
fixing bug in lowering of join. highlevel builder for 2Dstencil

Commit:
Highlevel builder for convolution supporting arbitrary kernels
- fixing padClamp requiring a padvalue

Commit:
Addressing comments on PR #8

- updating README
- description for all operations
- being more specific about argument names of operations
- fixing tests

Commit:
Major refactoring of loweringToImperative pass
- renaming of `ConT` and `AccT` to `lower` and `lowerAndStore`
- renaming loweringTarget flag on operations from `loop` to `scf`
- reflecting changes in tests and parser

Commit:
Refactoring for PR#8
- clarified parsing of Ops in Ops.cpp
- added parsing for result type of rise.embed
- adjusted tests

Commit:
clean up of lowerAndStore of ReduceSeq + MapSeq

michel-steuwer reviewed (approved):
+1
